lib/dynamo/dynamo/user.py

```diff
@@ -1,45 +1,93 @@
 import os
+from datetime import datetime, timezone
 from decimal import Decimal
 from os import environ
-from typing import Optional
 
 import botocore.exceptions
 
 from dynamo.util import DYNAMODB_RESOURCE
 
 
-def get_user(user_id: str) -> Optional[dict]:
-    table = DYNAMODB_RESOURCE.Table(environ['USERS_TABLE_NAME'])
-    response = table.get_item(Key={'user_id': user_id})
-    return response.get('Item')
+class DatabaseConditionException(Exception):
+    """Raised when a DynamoDB condition expression check fails."""
 
 
-def create_user(user_id: str) -> dict:
-    table = DYNAMODB_RESOURCE.Table(environ['USERS_TABLE_NAME'])
-    if get_user(user_id):
-        raise ValueError(f'user {user_id} already exists')
-    user = {'user_id': user_id, 'remaining_credits': Decimal(os.environ['DEFAULT_CREDITS_PER_USER'])}
-    table.put_item(Item=user)
+def get_or_create_user(user_id: str) -> dict:
+    current_month = _get_current_month()
+    default_credits = Decimal(os.environ['DEFAULT_CREDITS_PER_USER'])
+
+    users_table = DYNAMODB_RESOURCE.Table(environ['USERS_TABLE_NAME'])
+    user = users_table.get_item(Key={'user_id': user_id}).get('Item')
+
+    if user is not None:
+        user = _reset_credits_if_needed(
+            user=user,
+            default_credits=default_credits,
+            current_month=current_month,
+            users_table=users_table,
+        )
+    else:
+        user = _create_user(
+            user_id=user_id,
+            default_credits=default_credits,
+            current_month=current_month,
+            users_table=users_table,
+        )
+    return user
+
+
+def _get_current_month() -> str:
+    return datetime.now(tz=timezone.utc).strftime('%Y-%m')
+
+
+def _create_user(user_id: str, default_credits: Decimal, current_month: str, users_table) -> dict:
+    user = {'user_id': user_id, 'remaining_credits': default_credits, 'month_of_last_credits_reset': current_month}
+    try:
+        users_table.put_item(Item=user, ConditionExpression='attribute_not_exists(user_id)')
+    except botocore.exceptions.ClientError as e:
+        if e.response['Error']['Code'] == 'ConditionalCheckFailedException':
+            raise DatabaseConditionException(f'Failed to create user {user_id}')
+        raise
+    return user
+
+
+def _reset_credits_if_needed(user: dict, default_credits: Decimal, current_month: str, users_table) -> dict:
+    if (
+            os.environ['RESET_CREDITS_MONTHLY'] == 'yes'
+            and user['month_of_last_credits_reset'] < current_month  # noqa: W503
+            and user['remaining_credits'] is not None  # noqa: W503
+    ):
+        user['month_of_last_credits_reset'] = current_month
+        user['remaining_credits'] = default_credits
```

Comment:
We're missing the ability to give individual users a different monthly credit allotment (previously implemented via user['max_jobs_per_month']
```suggestion
        user['remaining_credits'] = user.get('credits_per_month', default_credits)
```